Refactor SSWU to use div_sqrt per Daira's suggestion. (#312)
cfrg/draft-irtf-cfrg-hash-to-curve: draft-irtf-cfrg-hash-to-curve.md

SWU method for any Weierstrass curve of the form given in {{weierstrass}}.
See {{simple-swu}} for information on the constants used in this mapping.

This optimized, straight-line procedure applies to any classes of curves

> and base field. The constant c1 depends on a fixed, non-square value h, which

Author: Oops. Good catch!

> is defined to be the smallest primitive element in the base field.

Reviewer: @daira is `h` required to be a primitive element, or just a non-square? If the latter, could we just use the (non-square) constant `Z` that is already defined?

Otherwise, we should somewhere document how to identify the smallest primitive element, or give a pointer to an existing procedure. (Sage could work, as in the sample code, but it's not obvious to me that `F.primitive_element()` is defined to return the smallest such element, so I worry a little about things breaking in the future...)

Reviewer 2: The algorithm that gives a square root of either s or h*s "for free" in a prime field F<sub>p</sub>, will generally require h to generate the order-2<sup>S</sup> subgroup where p-1 = 2<sup>S</sup> * T for odd T.

You could use Z = h (and I suggest that as a possibility [here](https://github.com/zcash/pasta/blob/f0f7068552a3565786cb338448cb58bc36a8314a/hashtocurve.sage#L269-L272)), but the I-D RECOMMENDS a specific Z that in general will not generate the order-T subgroup (and changing the procedure for finding the recommended Z now would not be compatible).

Reviewer: Thanks for the explanation.

One more question, just to be sure that I am understanding correctly: looks like as long as the `h` value used here and in `sqrt_ratio` are the same, there's no need for different implementations to use the same `h`. Is that right? If so, is the reason to recommend the smallest such value just efficiency?

Reviewer 2: Indeed different implementations need not use the same h. I'm not sure why the smallest value is recommended; that's not what my code used.

Author: > I'm not sure why the smallest value is recommended

Simply to remove choices for implementers. No other reason.

Reviewer: ```suggestion
is a parameter of the sqrt_ratio subroutine defined below.
```

Author: I'd be fine with that change as it still allows `S` to be computed deterministically.

Reviewer 2: I said above that the Z output by `find_z_sswu` wouldn't necessarily generate the order 2<sup>S</sup> subgroup, but actually it necessarily does, as pointed out in #317.

Output: (x, y), a point on E.

Constants:

> 1.  c1 = Z / h

Reviewer: ```suggestion
1.  c1 = Z / S
```

Reviewer: By the way, I have not checked this code carefully for correctness. I'll do that once the sqrt_ratio code is fixed.

Steps:
1.  tv1 = u^2
2.  tv1 = Z * tv1
3.  tv2 = tv1^2
4.  tv2 = tv2 + tv1
5.  tv3 = tv2 + 1
6.  tv3 = B * tv3
7.  tv4 = CMOV(Z, -tv2, tv2 != 0)
8.  tv4 = A * tv4
9.  tv2 = tv3^2
10. tv5 = tv4^2
11. tv5 = A * tv5
12. tv2 = tv2 + tv5
13. tv2 = tv2 * tv3
14. tv6 = tv4^3
15. tv5 = B * tv6
16. tv2 = tv2 + tv5
17.   x = tv1 * tv3
18. (is_gx1_square, y1) = sqrt_ratio(tv2, tv6)
19.   y = c1 * tv1
20.   y = y * u
21.   y = y * y1
22.   x = CMOV(x, tv3, is_gx1_square)
23.   y = CMOV(y, y1, is_gx1_square)
24. tv1 = u mod 2
25. tv2 = y mod 2
26.   y = CMOV(-y, y, tv1 == tv2)s
27.   x = x / tv4
28. return (x, y)
~~~

> This procedure depends on the following sqrt_ratio subroutine. It depends
> on the same value h defined above.

Reviewer: ```suggestion
The sqrt_ratio subroutine used by the above procedure is defined immediately below.
It depends on a constant S, which MUST be chosen to be a primitive element of the field F.
See below for more information on identifying an appropriate value for S.
```

I left out "non-square" here because a primitive element of GF(p^m) for odd p will always be non-square: p^m is odd, so 2 divides p^m - 1, so there's a subgroup of order 2; and a square cannot generate a subgroup of order 2.

~~~
sqrt_ratio(u, v)

Parameters:

> - F, a finite field of characteristic p and order q = p^m.

Reviewer: ```suggestion
- F, a finite field of characteristic p and order q = p^m.
- S, a primitive element of F.
```

Input: u and v, elements of F.

> Output: (b, y), where b is true and y = sqrt(u / v) if (u / v) is square in F,
>   and b is false and y = sqrt(h * (u / v)) otherwise.

Reviewer: ```suggestion
Output: (b, y), where b = True and y = sqrt(u / v) if (u / v) is square in F,
  and b = False and y = sqrt(S * (u / v)) otherwise.
```

Procedure:
1. m = 0
2. r = q - 1
3. while r % 2 == 0:
4.    r = r / 2
5.    m = m + 1

> 6. tv0 = h^r

Reviewer: ```suggestion
6. tv0 = S^r
```

7. tv1 = r + 1
8. tv1 = tv1 / 2

> 9. tv1 = h^tv1

Reviewer: ```suggestion
9. tv1 = S^tv1
```

10. tv2 = 2^m
11. tv2 = tv2 - 1

> 12. tv2 = D^tv2

Reviewer: ```suggestion
12. tv2 = v^tv2
```

The arguments are defined to be u,v not N,D.

All of the other pseudocode in the document was carefully checked by mechanically transforming the text from the document into the example code and then running the test suite. This new code should be checked likewise.

13. tv3 = tv2^2

> 14. tv3 = tv3 * D

Reviewer: ```suggestion
14. tv3 = tv3 * v
```

15. tv4 = r - 1
16. tv4 = tv4 / 2

> 17. tv5 = N * tv3

Reviewer: ```suggestion
17. tv5 = u * tv3
```

18. tv5 = tv5 ^ tv4
19. tv5 = tv5 * tv2

> 20. tv2 = tv5 * D # y

Reviewer: ```suggestion
20. tv2 = tv5 * v # y
```

> 21. tv3 = tv5 * N # z

Reviewer: ```suggestion
21. tv3 = tv5 * u # z
```

22. tv4 = tv3 * tv2 # t
23. tv5 = m-1
24. tv5 = 2^tv5
25. tv5 = tv4^tv5

> 26. isQR = CMOV(isQR, False, tv5 != 1)

Reviewer: `isQR` is use-before-def here. Need an initial `isQR = True` somewhere, I think.

27. tv3 = CMOV(tv3, tv3 * tv1, tv5 != 1)
28. tv4 = CMOV(tv4, tv4 * tv0, tv5 != 1)

> 29. for i in range(m, 1, -1):

Reviewer: ```suggestion
29. for i in (m, m - 1, ..., 2):
```

We do not use `range` in pseudocode elsewhere in the document.

30.    tv5 = i - 2
31.    tv5 = 2^tv5
32.    tv5 = tv4^tv5
33.    tv3 = CMOV(tv3, tv3 * tv0, tv5 != 1)
34.    tv4 = CMOV(tv4, tv4 * tv0, tv5 != 1)
35.    tv4 = CMOV(tv4, tv4 * tv0, tv5 != 1)
36.    tv0 = tv0 * tv0

> 37. return (isQR, tv3)

Reviewer: Below this point, need to add some very simple (probably Sage) pseudocode to show how to select a value for S. Maybe something like this:

```
The following Sage {{SAGE}} pseudocode outputs an appropriate S value for sqrt_ratio:

    # Arguments:
    # - F, a field object, e.g., F = GF(2^255 - 19)
    def find_S(F):
        return F.primitive_element()
```

(this is basically just aping the find_z_xxxx definitions later in the appendices).

Output: (x, y), a point on E.

Constants:
1.  c1 = Z / S

Steps:
1.  tv1 = u^2
2.  tv1 = Z * tv1
3.  tv2 = tv1^2
4.  tv2 = tv2 + tv1
5.  tv3 = tv2 + 1
6.  tv3 = B * tv3
7.  tv4 = CMOV(Z, -tv2, tv2 != 0)
8.  tv4 = A * tv4
9.  tv2 = tv3^2
10. tv5 = tv4^2
11. tv5 = A * tv5
12. tv2 = tv2 + tv5
13. tv2 = tv2 * tv3

> 14. tv6 = tv4^3

Reviewer 2: The RHS here is x1<sub>div</sub><sup>3</sup> in the notation used in the Zcash protocol spec (https://zips.z.cash/protocol/protocol.pdf#concretegrouphashpallasandvesta). We previously computed x1<sub>div</sub><sup>3</sup> as tv5 in step 10, and so there is no need to do a cubing here; a multiplication of (tv5 in step 10) * tv4 would be sufficient.

In general, I think that reusing temporary variables is unhelpful: it's more error-prone and harder for humans to understand, while the first thing a compiler is going to do anyway (regardless of platform, and platforms with few registers are not an exception) is to convert it to static single-assignment form.

Reviewer: Got the fix in the new PR.

I agree regarding temporary variables, but I'm going to focus on local fixes in my PR.

Author: Temporaries here are just the same as we do elsewhere in the doc. If we make a change here (which would be fine), we should at least be consistent.

> 11. tv5 = A * tv5
> 12. tv2 = tv2 + tv5
> 13. tv2 = tv2 * tv3
> 14. tv6 = tv4^3

Reviewer 2: ```suggestion
11. tv6 = tv5 * tv4
12. tv5 = A * tv5
13. tv2 = tv2 + tv5
14. tv2 = tv2 * tv3
```

15. tv5 = B * tv6
16. tv2 = tv2 + tv5
17.   x = tv1 * tv3
18. (is_gx1_square, y1) = sqrt_ratio(tv2, tv6)
19.   y = c1 * tv1
20.   y = y * u
21.   y = y * y1
22.   x = CMOV(x, tv3, is_gx1_square)
23.   y = CMOV(y, y1, is_gx1_square)
24.  e1 = sgn0(u) == sgn0(y)
25.   y = CMOV(-y, y, e1)
26.   x = x / tv4

> 27. return (x, y)

Reviewer 2: I have checked this against https://zips.z.cash/protocol/protocol.pdf#concretegrouphashpallasandvesta and it appears to be equivalent.

~~~

The sqrt_ratio subroutine used by the above procedure is defined immediately below.
It depends on a constant S, which MUST be chosen to be a primitive element of the field F.
See {{sswu-z-code}} for more information on identifying an appropriate value for S.

~~~
sqrt_ratio(u, v)

Parameters:
- F, a finite field of characteristic p and order q = p^m.
- S, a primitive element of F.

> Input: u and v, elements of F.

Reviewer 2: where v is nonzero (otherwise u / v is ill-defined). It may be useful to note that `map_to_curve_simple_swu` only calls it with v nonzero.

Output: (b, y), where b = True and y = sqrt(u / v) if (u / v) is square in F,
  and b = False and y = sqrt(S * (u / v)) otherwise.

Procedure:
1. isQR = True
2. m = 0
3. r = q - 1
4. while r % 2 == 0:
5.    r = r / 2
6.    m = m + 1
7. tv0 = S^r
8. tv1 = r + 1
9. tv1 = tv1 / 2
10. tv1 = S^tv1
11. tv2 = 2^m
12. tv2 = tv2 - 1
13. tv2 = v^tv2
14. tv3 = tv2^2
15. tv3 = tv3 * v
16. tv4 = r - 1
17. tv4 = tv4 / 2
18. tv5 = u * tv3
19. tv5 = tv5^tv4
20. tv5 = tv5 * tv2
21. tv2 = tv5 * v # y
22. tv3 = tv5 * u # z
23. tv4 = tv3 * tv2 # t
24. tv5 = m - 1
25. tv5 = 2^tv5
26. tv5 = tv4^tv5
27. isQR = CMOV(isQR, False, tv5 != 1)
28. tv3 = CMOV(tv3, tv3 * tv1, tv5 != 1)
29. tv4 = CMOV(tv4, tv4 * tv0, tv5 != 1)
30. for i in (m, m - 1, ..., 2):
31.    tv5 = i - 2
32.    tv5 = 2^tv5
33.    tv5 = tv4^tv5
34.    tv3 = CMOV(tv3, tv3 * tv0, tv5 != 1)
35.    tv4 = CMOV(tv4, tv4 * tv0, tv5 != 1)
36.    tv4 = CMOV(tv4, tv4 * tv0, tv5 != 1)
37.    tv0 = tv0 * tv0

> 38. return (isQR, tv3)

Reviewer 2: I have not checked this.

Output: (x, y), a point on E.

Constants:
1.  c1 = Z / S

Steps:
1.  tv1 = u^2
2.  tv1 = Z * tv1
3.  tv2 = tv1^2
4.  tv2 = tv2 + tv1
5.  tv3 = tv2 + 1
6.  tv3 = B * tv3
7.  tv4 = CMOV(Z, -tv2, tv2 != 0)
8.  tv4 = A * tv4
9.  tv2 = tv3^2
10. tv5 = tv4^2
11. tv5 = A * tv5
12. tv2 = tv2 + tv5
13. tv2 = tv2 * tv3
14. tv6 = tv4^3
15. tv5 = B * tv6
16. tv2 = tv2 + tv5
17.   x = tv1 * tv3
18. (is_gx1_square, y1) = sqrt_ratio(tv2, tv6)
19.   y = c1 * tv1
20.   y = y * u
21.   y = y * y1
22.   x = CMOV(x, tv3, is_gx1_square)
23.   y = CMOV(y, y1, is_gx1_square)

> 24. tv1 = u mod 2
> 25. tv2 = y mod 2
> 26.   y = CMOV(-y, y, tv1 == tv2)s
> 27.   x = x / tv4
> 28. return (x, y)

Reviewer: ```suggestion
24.  e1 = sgn0(u) == sgn0(y)
25.   y = CMOV(-y, y, e1)
26.   x = x / tv4
27. return (x, y)
```

Should use the sgn0 abstraction: `{u,y} mod 2` isn't correct for extension fields.

~~~

The sqrt_ratio subroutine used by the above procedure is defined immediately below.
It depends on a constant S, which MUST be chosen to be a primitive element of the field F.

> See below for more information on identifying an appropriate value for S.

Author: I don't feel strongly. Keeping it where it is now has the advantage of not sprinkling Sage everywhere.
